### PR TITLE
Expose some fields of GDBusNodeInfo

### PR DESCRIPTION
**Pull request by abergmeier**
This exposes fields as functions:
- path
- interfaces
- nodes

### CODE REVIEW

File: gio/src/dbus_node_info.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        let c_obj = self.to_glib_none().0;
+        let path = unsafe { (*c_obj).path };
+        if path.is_null() {
+            return "";
+        }
+        let c_str = unsafe { CStr::from_ptr(path) };
+        c_str.to_str().unwrap()
+    }
+
+    pub fn interfaces(self) -> Vec<DBusInterfaceInfo> {
```
Comment (bilelmoussaoui) [inline gio/src/dbus_node_info.rs:20]: Do you think you can also expose the fields of the other DBus*Info types?

**Comment by abergmeier** [comment]
Not totally sure whether ownership semantics are correct.

File: gio/src/dbus_node_info.rs (hunk 1 above)
Comment (abergmeier) [reply gio/src/dbus_node_info.rs:20]: I could but I currently have no use case for them. And I feel kind of bad adding too much code that is not backed up by tests.

File: gio/src/dbus_node_info.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+use std::ffi::CStr;
+
+use glib::translate::ToGlibPtr;
+
+use crate::{DBusInterfaceInfo, DBusNodeInfo};
+
+impl DBusNodeInfo {
+    pub fn path(self) -> &'static str {
+        let c_obj = self.to_glib_none().0;
```
Comment (sdroege) [inline gio/src/dbus_node_info.rs:11]: I'd make this one big `unsafe` block in this function as it's all related really

File: gio/src/dbus_node_info.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::ffi::CStr;
+
+use glib::translate::ToGlibPtr;
+
+use crate::{DBusInterfaceInfo, DBusNodeInfo};
+
+impl DBusNodeInfo {
+    pub fn path(self) -> &'static str {
```
Comment (sdroege) [inline gio/src/dbus_node_info.rs:10]:
```suggestion
    pub fn path<'a>(&'a self) -> &'a str {
```

probably. Can this be `NULL`, if so it should probably be

```suggestion
    pub fn path<'a>(&'a self) -> Option<&'a str> {
```

File: gio/src/dbus_node_info.rs (hunk 1 above)
Comment (sdroege) [inline gio/src/dbus_node_info.rs:20]:
```suggestion
    pub fn interfaces<'a>(&'a self) -> &'a [DBusInterfaceInfo] {
```

via https://gtk-rs.org/gtk-rs-core/git/docs/glib/collections/struct.PtrSlice.html#method.from_glib_borrow_num should also work. We should add a non-`_num` version of this.

File: gio/src/dbus_node_info.rs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        let c_obj = self.to_glib_none().0;
+        let c_ii = unsafe { (*c_obj).interfaces };
+        if c_ii.is_null() {
+            return Vec::new();
+        }
+
+        unsafe { glib::translate::FromGlibPtrContainer::from_glib_none(c_ii) }
+    }
+
+    pub fn nodes(self) -> Vec<DBusNodeInfo> {
```
Comment (sdroege) [inline gio/src/dbus_node_info.rs:30]:
```suggestion
    pub fn nodes<'a>(&'a self) -> &'a [DBusNodeInfo] {
```

as above

File: gio/src/dbus_node_info.rs (hunk 1 above)
Comment (abergmeier) [reply gio/src/dbus_node_info.rs]: Done

File: gio/src/dbus_node_info.rs (hunk 4 above)
Comment (abergmeier) [reply gio/src/dbus_node_info.rs]: Done

File: gio/src/dbus_node_info.rs (hunk 2 above)
Comment (abergmeier) [reply gio/src/dbus_node_info.rs:11]: Done

File: gio/src/dbus_node_info.rs (hunk 3 above)
Comment (abergmeier) [reply gio/src/dbus_node_info.rs]: Done

File: gio/src/dbus_node_info.rs (hunk 4 above)
Comment (abergmeier) [reply gio/src/dbus_node_info.rs]: Doing this fails the CI: https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

Comment (sdroege) [reply gio/src/dbus_node_info.rs]: Oh yes you can omit all the `'a` here, sorry

Comment (abergmeier) [reply gio/src/dbus_node_info.rs]: Done

File: gio/src/dbus_node_info.rs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            let path = (*c_obj).path;
+            if path.is_null() {
+                return None;
+            }
+            let c_str = CStr::from_ptr(path);
+            Some(c_str.to_str().unwrap())
+        }
+    }
+
+    pub fn interfaces(self) -> &'static [DBusInterfaceInfo] {
```
Comment (sdroege) [inline gio/src/dbus_node_info.rs:22]:
```suggestion
    pub fn interfaces(&self) -> &[DBusInterfaceInfo] {
```

File: gio/src/dbus_node_info.rs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            let c_ii = (*c_obj).interfaces;
+            if c_ii.is_null() {
+                return &[];
+            }
+
+            glib::collections::PtrSlice::from_glib_borrow(c_ii)
+        }
+    }
+
+    pub fn nodes(self) -> &'static [DBusNodeInfo] {
```
Comment (sdroege) [inline gio/src/dbus_node_info.rs:34]:
```suggestion
    pub fn nodes(&self) -> &[DBusNodeInfo] {
```

File: gio/src/dbus_node_info.rs (hunk 5 above)
Comment (abergmeier) [reply gio/src/dbus_node_info.rs]: Done

File: gio/src/dbus_node_info.rs (hunk 6 above)
Comment (abergmeier) [reply gio/src/dbus_node_info.rs]: Done

**Review by sdroege — APPROVED** [review]

**Comment by sdroege** [comment]
Thanks!

**Pull request merged**